Source/EasyNetQ/PullingConsumer.cs

```diff
@@ -0,0 +1,435 @@
+        private readonly ulong messagesCount;
+
+
+        /// <summary>
+        ///     Represents a result when no message is available
+        /// </summary>
+        public static PullResult<T> NotAvailable { get; } = new PullResult<T>(false, 0, null, null);
+
+        /// <summary>
+        ///     Represents a result when a message is available
+        /// </summary>
+        public static PullResult<T> Available(
+            ulong messagesCount, MessageReceivedInfo receivedInfo, IMessage<T> message
+        )
+        {
+            return new PullResult<T>(true, messagesCount, receivedInfo, message);
+        }
+
+        private PullResult(bool isAvailable, ulong messagesCount, MessageReceivedInfo receivedInfo, IMessage<T> message)
+        {
+            this.isAvailable = isAvailable;
+            this.messagesCount = messagesCount;
+            this.receivedInfo = receivedInfo;
+            this.message = message;
+        }
+
+        /// <summary>
+        ///     True is a message is available
+        /// </summary>
+        public bool IsAvailable => isAvailable;
+
+        /// <summary>
+        ///     Returns remained messages count if the message is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public ulong RemainedMessagesCount
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return messagesCount;
+            }
+        }
+
+        /// <summary>
+        ///     Returns received info if the message is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public MessageReceivedInfo ReceivedInfo
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return receivedInfo;
+            }
+        }
+
+        /// <summary>
+        ///     Returns message if it is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public IMessage<T> Message
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return message;
+            }
+        }
+    }
+
+    /// <summary>
+    ///     Allows to receive messages by pulling them one by one
+    /// </summary>
+    public interface IPullingConsumer : IDisposable
+    {
+        /// <summary>
+        ///     Receives a single message
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns></returns>
+        Task<PullResult> PullAsync(CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///     Acknowledges one or more messages
```

Comment:
> one _or more_ messages

I think this requires further clarification about multiple acks. What needs to be in deliveryTag? Tag of the last pulled message or ... ?

Reply:
Maximum of delivery tags from messages